Documentation/ceph-object-bucket-claim.md

```diff
@@ -0,0 +1,242 @@
+1. `additionalConfig` gives providers a location to set proprietary config values (tenant, namespace...).
+The value is a list of 1 or more key-value pairs.
+
+##### OBC Custom Resource (after update by lib)
+```yaml
+apiVersion: objectbucket.io/v1alpha1
+kind: ObjectBucketClaim
+metadata:
+  finalizers: [1]
+  - objectbucket.io/finalizer
+  labels: [2]
+    bucket-provisioner: aws-s3.io-bucket [3]
+spec:
+  bucketName: photo-booth-62PrQ [4]
+  objectBucketRef: objectReference{} [5]
+  configMapRef: objectReference{} [6]
+  secretRef: objectReference{} [7]
+status:
+  phase: {"Pending", "Bound", "Released", "Failed"} [8]
+```
+1. the `finalizer` added by the library, the name is a constant.
+1. the library adds a `label` (seen here) but each provisioner can
+   supply their own `labels`.
+1. the label value shown is the name of the provisioner.
+   In this example the provisioner name is `aws-s3.io/bucket`.
+1. the generated, unique `bucket name` for the new bucket.
+1. `objectReference` to the generated `OB`.
+1. `objectReference` to the generated `ConfigMap`.
+1. `objectReference` to the generated `Secret`.
+1. phases of bucket creation:
+    - _Pending_: the operator is processing the request
+    - _Bound_: the operator finished processing the request and linked the OBC and OB
+    - _Released_: the OB has been deleted, leaving the OBC unclaimed but unavailable.
+    - _Failed_: not currently set.
+
+##### Generated Secret (sample for rook-ceph provider)
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: ceph-bucket
+  namespace: rook-ceph
+  finalizers:
+  - objectbucket.io/finalizer
+  labels:
+    bucket-provisioner: aws-s3.io-bucket
+  ownerReferences:
+  - name: ceph-bucket [1]
+    ...
+type: Opaque
+data:
+  ACCESS_KEY_ID: BASE64_ENCODED-1
+  SECRET_ACCESS_KEY: BASE64_ENCODED-2
+  ... [2]
+```
+1. ownerReference makes this secret a child of the originating OBC for clean up purposes.
+1. ACCESS_KEY_ID and SECRET_ACCESS_KEY are the only secret keys defined by the library.
+Provisioners are able to cause the lib to create additional keys by returning  the `AdditionalSecretConfig` field.
+**Note:** the library will create the Secret using `stringData:` and let the Secret API base64 encode the values.
+Eg:
+```
+stringData:
+  endpoint: |-
+    ACCESS_KEY_ID: NON-BASE64-STRING
+    SECRET_ACCESS_KEY: NON-BASE64-STRING
+```
+
+##### Generated ConfigMap (sample for rook-ceph provider)
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: ceph-bucket
+  namespace: rook-ceph
+  finalizers:
+  - objectbucket.io/finalizer
+  labels:
+    bucket-provisioner: aws-s3.io-bucket
+  ownerReferences:
+  - name: rook-ceph
+    ...
+data:
+  BUCKET_HOST: http://my-store [1]
+  BUCKET_PORT: 80 [2]
+  BUCKET_NAME: photo-booth-62PrQ [3]
+  BUCKET_REGION: us-west-1
+  ... [4]
+```
+1. host URL.
+1. host port.
+1. unique bucket name.
+1. the above data keys are defined by the library.
```

Comment:
~~1. the above data keys are defined by the library.~~   don't think this is important and I don't want to use the term _library_.